dask/dataframe/io.py

```diff
@@ -637,23 +637,37 @@ def to_castra(df, fn=None, categories=None, sorted_index_column=None,
         return delayed([Delayed(key, [dsk]) for key in keys])
 
 
-def to_csv(df, filename, compression=None, get=None, compute=True, **kwargs):
+def to_csv(df, filename, name_function=None, compression=None, get=None, compute=True, **kwargs):
     if compression:
         raise NotImplementedError("Writing compressed csv files not supported")
     name = 'to-csv-' + uuid.uuid1().hex
 
+    kwargs2 = kwargs.copy()
+
+    if name_function is None:
+        name_function = build_name_function(df.npartitions - 1)
+
+    if '*' in filename:
+        formatted_names = [name_function(i) for i in range(df.npartitions)]
+        if formatted_names != sorted(formatted_names):
+            warn("In order to preserve order between partitions "
+                 "name_function must preserve the order of its input")
+
+        if filename.count('*') > 1:
+            raise ValueError("A maximum of one asterisk is accepted in filename")
+    else:
+        kwargs2['mode'] = 'a'
+        kwargs2['header'] = False
+
     dsk = dict()
     dsk[(name, 0)] = (lambda df, fn, kwargs: df.to_csv(fn, **kwargs),
-                        (df._name, 0), filename, kwargs)
-
-    kwargs2 = kwargs.copy()
-    kwargs2['mode'] = 'a'
-    kwargs2['header'] = False
+                        (df._name, 0), filename.replace('*', name_function(0)), kwargs)
 
     for i in range(1, df.npartitions):
+        filename_i = filename.replace('*', name_function(i))
         dsk[(name, i)] = (_link, (name, i - 1),
```

Comment:
> We aim to keep the order of partitions, and that means one must follow the other, in the same way they are partitioned.

Yup but I guess that keeping the order of partitions when writing to multiple files consists on naming them using a sorted scheme like `part-001.csv`, `part-002.csv`, etc., where each part file corresponds to a partition, but you can write all these part files at the same time, you don't need the first one to be written before starting with the second one, right?

This is in constrast with the single file mode, where you need the each part to be written (especially the first one, that contains the headers) before being able to write the following one to keep the order.